app-server/database.js

```diff
@@ -0,0 +1,771 @@
+          promises2.push(this._createHelper_addShowingDate(conn, showId, opening));
+
+          if (!isSingleEvent) {
+            promises2.push(this._createHelper_addShowingDate(conn, showId, closing));
+          }
+          if (scriptId) {
+            promises2.push(this._createHelper_makeStagingScripted(conn, stagingId, scriptId));
+          }
+          if (spaceId) {
+            promises2.push(this._createHelper_setPerformanceSpaceForShow(conn, showId, spaceId));
+          }
+
+          Promise.all(promises2).then(
+            () => {
+              resolve({showId});
+              conn.destroy();
+            }
+          ).catch(
+            reason => {
+              reject(reason);
+              conn.destroy();
+            }
+          );
+        }
+      ).catch(
+        reason => {
+          reject(reason);
+          conn.destroy();
+        }
+      );
+
+    });
+  }
+  static _createHelper_insertShow(conn, orgId) {
+    return new Promise((resolve, reject) => {
+      var query = "insert into `show` (producer_id) values (?)";
+      conn.query(query, [orgId], (err, result) => {
+        if (err) {
+          reject(err);
+        } else {
+          resolve(result.insertId);
+        }
+      });
+    });
+  }
+  static _createHelper_insertStaging(conn, inputProduction) {
+    return new Promise((resolve, reject) => {
+      var query, params;
+      if (inputProduction.isScripted) {
+        query = "insert into staging () values ()";
+        params = [];
+      } else {
+        var {stagingTitle, description} = inputProduction;
+        query = "insert into staging (name, staging_notes) values (?, ?)";
+        params = [stagingTitle, description];
+      }
+      conn.query(query, params, (err, result) => {
+        if (err) {
+          reject(err);
+        } else {
+          resolve(result.insertId);
+        }
+      });
+    });
+  }
+  static _createHelper_insertScript(conn, inputProduction) {
+    return new Promise((resolve, reject) => {
+      var {scriptTitle, synopsis} = inputProduction;
+      var query = "insert into script (name, synopsis) values (?, ?)";
+      conn.query(query, [scriptTitle, synopsis], (err, result) => {
+        if (err) {
+          reject(err);
+        } else {
+          resolve(result.insertId);
+        }
+      });
+    });
+  }
+  static _createHelper_createShowOrderWithStaging(conn, showId, stagingId) {
+    return new Promise((resolve, reject) => {
+      var query = "insert into show_order (show_id, staging_id, `order`) values (?, ?, ?)";
+      conn.query(query, [showId, stagingId, 1], (err, result) => {
+        if (err) {
+          reject(err);
+        } else {
+          resolve(result.insertId);
+        }
+      });
+    });
+  }
+  // "date" is a timestamp using milliseconds.
+  static _createHelper_addShowingDate(conn, showId, date) {
+    return new Promise((resolve, reject) => {
```

Comment:
In general, you should only use the `new Promise` constructor when necessary. It's a lot easier if you can keep more of your logic in promise-land, which are more composable and make it easier to avoid nesting.

One thing that might help is creating a couple helpers that make composing the promises easier (probably bad example code below):

```javascript
function handleInsertResult(err, result) {
  if (err) return Promise.reject(err);
  else return result.insertId;
}

function queryConnection(connection, query, args) {
  return new Promise(resolve => {
    connection.query(query, args, resolve);
  });
}
```

Then this can be used:

```javascript
let query = "insert into scheduled_showing (show_id, showtime) VALUES (?, FROM_UNIXTIME(?))";
queryConnection(connection, query, [stagingId, scriptId])
  .then(handleInsertResult)
  .then(console.log.bind(console))
  .catch(console.error.bind(console));
```

This moves you as quickly as possible out of manually creating promises, which I find a lot easier to think about.